dev/test/bulk-writer.ts

```diff
@@ -477,6 +482,146 @@ describe('BulkWriter', () => {
+    const bulkWriter = await instantiateInstance([
+      {
+        request: createRequest([setOp('doc', 'bar')]),
+        response: failedResponse(Status.FAILED_PRECONDITION),
+      },
+      {
+        request: createRequest([setOp('doc', 'bar')]),
+        response: successResponse(1),
+      },
+    ]);
+    let op: BulkWriterOperation | undefined = undefined;
+    bulkWriter.onError(async (error, operation) => {
+      op = operation;
+    });
+    bulkWriter.set(firestore.doc('collectionId/doc'), {foo: 'bar'});
+    await bulkWriter.flush();
+    expect(op).to.not.be.undefined;
+    let writeResult: WriteResult;
+    bulkWriter.retry(op!).then(res => {
+      writeResult = res;
+    });
+    return bulkWriter.close().then(() => {
+      expect(writeResult.writeTime.isEqual(new Timestamp(1, 0))).to.be.true;
+    });
+  });
+
+  it('can retry failed update operations with retry()', async () => {
+    const bulkWriter = await instantiateInstance([
+      {
+        request: createRequest([updateOp('doc', 'bar')]),
+        response: failedResponse(Status.FAILED_PRECONDITION),
+      },
+      {
+        request: createRequest([updateOp('doc', 'bar')]),
+        response: successResponse(1),
+      },
+    ]);
+    let op: BulkWriterOperation | undefined = undefined;
+    bulkWriter.onError(async (error, operation) => {
+      op = operation;
+    });
+    bulkWriter.update(firestore.doc('collectionId/doc'), {foo: 'bar'});
+    await bulkWriter.flush();
+    expect(op).to.not.be.undefined;
+    let writeResult: WriteResult;
+    bulkWriter.retry(op!).then(res => {
+      writeResult = res;
+    });
+    return bulkWriter.close().then(() => {
+      expect(writeResult.writeTime.isEqual(new Timestamp(1, 0))).to.be.true;
+    });
+  });
+
+  it('can retry failed delete operations with retry()', async () => {
+    const bulkWriter = await instantiateInstance([
+      {
+        request: createRequest([deleteOp('doc')]),
+        response: failedResponse(Status.FAILED_PRECONDITION),
+      },
+      {
+        request: createRequest([deleteOp('doc')]),
+        response: successResponse(1),
+      },
+    ]);
+    let op: BulkWriterOperation | undefined = undefined;
+    bulkWriter.onError(async (error, operation) => {
+      op = operation;
+    });
+    bulkWriter.delete(firestore.doc('collectionId/doc'));
+    await bulkWriter.flush();
+    expect(op).to.not.be.undefined;
+    let writeResult: WriteResult;
+    bulkWriter.retry(op!).then(res => {
+      writeResult = res;
+    });
+    return bulkWriter.close().then(() => {
+      expect(writeResult.writeTime.isEqual(new Timestamp(1, 0))).to.be.true;
+    });
+  });
+
+  it('retried operations are enqueued after other operations', async () => {
+    const bulkWriter = await instantiateInstance([
+      {
+        request: createRequest([setOp('doc1', 'bar'), updateOp('doc2', 'bar')]),
+        response: mergeResponses([
+          failedResponse(Status.FAILED_PRECONDITION),
```

Comment:
How about `Internal`? `FAILED_PRECONDITION` is a bit strange since the retry should also fail.